Enable testing without root privileges

wferi opened the pull request:
If /proc/sys/net/core/[rw]mem_max are set to at least 8388608
(KNET_RING_RCVBUFF), the tests don't need root privileges.

This is admittedly sketchy, for example I left in the message stating the need for root privileges.
Maybe the error reporting could be made more detailed at well.

By the way: are large socket buffers critical for knet? Would it make any sense to run/test without them?

wferi committed:
Don't require root privileges unless necessary

On Linux, if /proc/sys/net/core/[rw]mem_max are set to at least
8388608 (KNET_RING_RCVBUFF), setting the socket buffer sizes
doesn't require root privileges.

FreeBSD uses the kern.ipc.maxsockbuf sysctl MIB variable for
capping user buffer requests.

Linux doubles the requested amount for administrative overhead,
but FreeBSD does not, so we can't be too strict when checking
the results.

wferi commented:
On a second thought, it may be better to switch the fallback order around. But first let's agree on the general idea and clear up the above questions.

fabbione commented:
Hi Feri, I think being able to run the test suite as non-root is ok. honestly I didn´t look at the patch too closely but I would like some time to think on how to do get it done. Originally I was considering to look for CAPABILITIES and alike, but I am happy to explore other options once 1.0 is out of the door.

wferi commented:
The tests run fine as non-root if `/proc/sys/net/core/[rw]mem_max` are set to at least 8388608 and we don't bail out if `SO_RCVBUFFORCE` or `SO_SNDBUFFORCE` fail (unfortunately they fail even when the required buffer lengths are below the configured maxima and could be set by plain `SO_RCVBUF` and `SO_SNDBUF`). I guess it means knet itself runs fine as non-root in such an environment with this change. The question is whether knet should start up printing a warning if it can't set the usual buffer lengths (lacking CAP_NET_ADMIN or a generous `/proc` setup) or simply refuse to initialize at all.

Anyway, I'll shortly push another commit switching the operations around.

wferi commented:
These additional patches don't change functionality, just demonstrate different approaches.

wferi commented:
Hmm, looks like FreeBSD does not double the buffer like Linux does. No big deal.

fabbione commented:
Hey Feri, sorry I am not ignoring this patch. We found a couple of rather nasty bugs and we are working to get them fixed before 0.9.

wferi commented:
Sure, take your time fixing bugs. Meanwhile I rebased and squashed the series. The last FreeBSD failure was rather mysterious, I wonder if it happens again...

fabbione commented:
The last failures are all timeout related to those issues we are fixing :-)

fabbione commented:
Feri, i´ll try to take a look at it over the weekend. We managed to fix one of the major bugs, now we need to look into the other one. As for the BSD failures, I noticed some are related to the timeout issue we just fixed, others appear to be specific to this patch, but it shouldn´t be hard to fix.

wferi committed:
Remove the need_root() function

The point is not having root privileges, but having high enough socket
buffer caps.  Very few of the test can run without it.

wferi committed:
Detailed error reporting about socket buffer sizes

wferi commented:
Something like this? Yell if you prefer it squashed or anything.
